### PR TITLE
mk: add .so name versioning, resolves #32

### LINKED ISSUE #32: Upstream .so name versioning?
(Basically the same as at https://github.com/creytiv/re/issues/249, but it seems that this is the new home of re)

As of writing, there doesn't seem to be proper upstream .so name versioning in re. Yes, I'm aware there is something like `LIB_SUFFIX=.so.0`, but this unfortunately has nothing to do with proper .so name versioning, as the SONAME is still `libre.so` rather `libre.so.0`. Yes, even this can be worked around by passing some arguments like `SH_LFLAGS="-shared -Wl,-soname,libre.so.0"`, but this still doesn't let downstreams know whether the old and the new version are ABI compatible or not, thus this should be IMHO really handled by upstream.

Quoting from section "SONAME handling" from https://docs.fedoraproject.org/en-US/packaging-guidelines/#_downstream_so_name_versioning:

> The SONAME field is written to the shared object by linker, using (at least in case of `ld`) the `-soname SONAME` flags. This can be passed as an option to gcc like this:
>
> ```
> $ gcc $CFLAGS -Wl,-soname,libfoo.so.0.n -o libfoo.so.0.n
> ```
> 
> If you want to check if the SONAME field is set and what value it has, use the `objdump` command (from `binutils`):
>
> ```
> $ objdump -p /path/to/libfoo.so.0.n | grep 'SONAME'
> ```

Please note that this issue is also meant as very kind try to convince you as upstream to start versioning your .so library (as our Fedora Packaging Guidelines request).

### PR DESCRIPTION
**Pull request by sreimers**
@robert-scheck Can you check if soname handling fits your needs/requirements?

@alfredh I saw that you added and reverted a similar change some time ago: https://github.com/creytiv/re/commit/1d191277024abd512491fdcbc270649565bf352d

Are there any problems I should be aware of or i should test?

### CODE REVIEW

File: mk/re.mk [hunk 1]
```diff
@@ -233,7 +233,7 @@ ifeq ($(OS),linux)
 	CFLAGS		+= -fPIC -DLINUX
 	LIBS		+= -ldl
 	LFLAGS		+= -fPIC
-	SH_LFLAGS	+= -shared
+	SH_LFLAGS	+= -shared -Wl,-soname,libre.so.$(VER_ABI)
```
Comment (alfredh) [inline mk/re.mk:236]: re.mk is a shared makefile that can also be used by other projects.

**Comment by alfredh** [comment]
after building this branch I get:

```
$ objdump -p libre.so | grep 'SONAME'
  SONAME               libre.so.11
```

is it correct that the number after so should be 11 ?

File: mk/re.mk (hunk 1 above)
Comment (sreimers) [reply mk/re.mk]: I overlooked that, is fixed now.

**Comment by sreimers** [comment]
> is it correct that the number after so should be 11 ?

Yes it's currently `VER_ABI   := $(VER_MAJOR)$(VER_MINOR)`, but we can start by `libre.so.0` and count ABI different.

My thought was a new Feature release, increments ABI too. but we can also specify a stricter ABI compatibility
and setup some tests.

**Comment by alfredh** [comment]
I am not sure about this PR.

 I guess @robert-scheck should test and review it.

I sometimes look at the Opus project for inspiration. Here is similar info
for libopus from Debian 10:


```
$ dpkg -l | grep libopus0
ii  libopus0:amd64                         1.3-1                               amd64        Opus codec runtime library
ii  libopus0:i386                          1.3-1                               i386         Opus codec runtime library

$ l /usr/lib/x86_64-linux-gnu/libopus.*
-rw-r--r-- 1 root root 701100 Dec 17  2018 /usr/lib/x86_64-linux-gnu/libopus.a
lrwxrwxrwx 1 root root     16 Dec 17  2018 /usr/lib/x86_64-linux-gnu/libopus.so -> libopus.so.0.7.0
lrwxrwxrwx 1 root root     16 Dec 17  2018 /usr/lib/x86_64-linux-gnu/libopus.so.0 -> libopus.so.0.7.0
-rw-r--r-- 1 root root 370600 Dec 17  2018 /usr/lib/x86_64-linux-gnu/libopus.so.0.7.0

$ objdump -p /usr/lib/x86_64-linux-gnu/libopus.so | grep SONAME
  SONAME               libopus.so.0
```

I dont understand why libopus0 has version 1.3 ...?

perhaps we should also ask @jonassmedegaard from the Debian project for input.


Alfred

**Comment by jonassmedegaard** [comment]
The Debian package containing the runtime library has the major part of the SONAME in its package *name* and has the 
upstream opus release version in its package version.

1.3 is version of upstream opus release
1.3-1 is version of Debian package (upstream version + Debian packaging revision)
0.7.0 is SONAME

Hope that helps.

**Comment by sreimers** [comment]
I updated the PR to count SONAME ABI Version independent and starting with `libre.so.1` (prepared for next release).

**Comment by sreimers** [comment]
After some thought and research, I would prefer this versioning:
```
/usr/lib/libre.so -> libre.so.1.1.0
/usr/lib/libre.so.1 -> libre.so.1.1.0
/usr/lib/libre.so.1.1.0
```
```
$ objdump -p /usr/lib/libre.so | grep SONAME
  SONAME               libre.so.1
```  
This corresponds to [semantic versioning](https://semver.org/spec/v2.0.0.html).
Breaking the ABI/API requires bumping the Major Version.

Sorry for the back and forth. But I personally use Libre only as static cross-platform builds
and had no preference for dynamic builds and soname versioning. Different projects uses
different approaches which is bit confusing, too.

The "/usr/lib/libre.so.1 -> libre.so.1.1.0" linking is done by ldconfig (soname field). As
mentioned here:

https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html

**Comment by jonassmedegaard** [comment]
This text seems to decently clarify the distinction between project versions, APIs and ABIs, in case that's a help to some following along here: https://dcreager.net/shared-library-versions/

**Comment by sreimers** [comment]
Ok the author suggest to use a separate ABI shared libary version info, the math is the following:
```
This flag accepts an argument of the form ‘current[:revision[:age]]’. So, passing -version-info 3:12:1 sets current to 3, revision to 12, and age to 1.

$ABI_MAJOR = $current - $age
versuffix=$major.$age.$revision
SOVERSION=$major
```

- Programs using the previous version may use the new version as drop-in replacement, and programs using the new version can also work with the previous one. In other words, no recompiling nor relinking is needed. In this case, bump revision only, don’t touch current nor age.
- Programs using the previous version may use the new version as drop-in replacement, but programs using the new version may use APIs not present in the previous one. In other words, a program linking against the new version may fail with “unresolved symbols” if linking against the old version at runtime: set revision to 0, bump current and age.
- Programs may need to be changed, recompiled, and relinked in order to use the new version. Bump current, set revision and age to 0.

https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html

**Pull request merged**